### DISCUSSION TITLE
How can I act on an instance reference ?

### DISCUSSION THREAD

**Discussion by jugglingjsons** [quaadgras/graphics.gd#27] [Q&A] [answered]
Is it possible to directly call native methods on instances of Objects I created via gd ? 

Sample:

```go
func (m *Main) createBoid(id int) {
	mob, ok := classdb.As[gd.RigidBody2D](m.MobScene.Instantiate())
	if !ok {
		fmt.Println("failed to cast!")
		return
	}
```

I would Ideally want to have a reference to the mob, that I could call .setVelocity / .setPosition on.
Would that be possible ?

For context when I later use:

```go
	b.gdObj.AsNode2D().SetPosition(position)
	b.gdObj.SetLinearVelocity(velocity)
```

it ends up exploding with: `panic: runtime.link/mmm error: use after free`

I may need to provide a reproducible example, I am using rwMutex on the boids, likely I screw up access in there, and it's not about the method linking at all. Will report back :)

**Comment by Splizard** [comment]
I imagine you are storing the mob somewhere, in order to do this with a Godot reference, you will need to pin the lifetime of the mob to the class it is contained in. I am exploring a design on how to make this more intuiative.
(See https://github.com/grow-graphics/gd/issues/25#issuecomment-2157824522)

Note that since the instance inherits `gd.Node` adding the node as a child to another node will transfer ownership over to Godot (hence use-after 'free'), you may find it easier to add the 'mob' to the scene tree and then keep a `gd.NodePath` to it so that you can retrieve it at a later point.

A good heuristic to keep things simpler when working with Godot references in Go, always try to use references temporarily (within your method call) before you return back to Godot, if you need to hold onto the reference for a longer time, store the reference somewhere inside a Godot managed class (or use a Godot reference like `gd.NodePath` or `gd.ObjectID`), so that you can retrieve it later. This will help avoid issues with `grow.graphics/gd` semi-automatic memory management.

**Reply by jugglingjsons** [reply]
I don't seem to understand how to get it working.
I've prepared a sample below.

```go
package main

import (
	"fmt"
	"math/rand"

	"graphics.gd/classdb"
	"graphics.gd/classdb/Node"
	"graphics.gd/classdb/PackedScene"
	"graphics.gd/classdb/RigidBody2D"
	"graphics.gd/startup"
	"graphics.gd/variant/Angle"
	"graphics.gd/variant/Float"
	"graphics.gd/variant/NodePath"
	"graphics.gd/variant/Vector2"
	"grow.graphics/gd"
)

type Main struct {
	classdb.Extension[Main, Node.Instance] `gd:"Boids"`
	MobScene                               PackedScene.Instance
	boidsGD                                []NodePath.String
}

const (
	boidCount  = 10
	viewRadius = 25
	adjRate    = 0.015
)

const (
	screenHeight = 320 * 4
	screenWidth  = 640 * 4
)

var boidMap [screenWidth + 1][screenHeight*4 + 1]int

func (m *Main) Ready() {
	for i := 0; i < boidCount; i++ {

		instance := Node.Instance(m.MobScene.Instantiate())
		mob, ok := classdb.As[RigidBody2D.Instance](instance)
		if !ok {
			fmt.Println("failed to cast!")
			return
		}

		positionGD := Vector2.XY{}
		positionGD.X = Float.Random() * screenWidth
		positionGD.Y = Float.Random() * screenHeight
		velocityGD := Vector2.New((Float.Random()*2)-1, (Float.Random()*2)-1)

		// Add some randomness to the direction.
		direction := Angle.Radians(Float.RandomBetween(Float.X(-Angle.Pi/4), Float.X(Angle.Pi/4)))

		mob.AsNode2D().SetPosition(positionGD)
		mob.AsNode2D().SetRotation(Float.X(direction))

		mob.SetLinearVelocity(Vector2.Rotated(velocityGD, direction))

		m.Super().AddChild(instance)
		fmt.Println(instance.GetPath())
		m.boidsGD = append(m.boidsGD, instance.GetPath())
	}
	fmt.Println("boids created", m.boidsGD)
}

var velocity = gd.Vector2{}

func (m *Main) Process(godot gd.Context, delta gd.Float) {
	for i, boidPath := range m.boidsGD {

		node := m.Super().GetNode(boidPath)

		idx := i
		velocity.SetX(velocity.X() + rand.Float64() - 0.5)
		velocity.SetY(velocity.Y() + rand.Float64() - 0.5)

		mob, ok := gd.As[gd.RigidBody2D](godot, node)
		if !ok {
			fmt.Println("failed to cast!")
			return
		}

		fmt.Print("casted ", idx, " ", mob)
		// mob.SetPosition(postion)
		mob.SetLinearVelocity(velocity)
	}
}

func main() {
	classdb.Register[Main]()
	classdb.Register[Mob]()
	startup.Engine()
}
```

It also explodes with the same error.

**Reply by Splizard** [reply, marked as answer]
Thanks, looks like you have found a bit of an issue here, as you need the node to be alive in Go in order to get the nodepath, I've added a function to `gd` (`gd.AddChild`) that will workaround this, it will add the child and return the resulting NodePath (I've tested this and it looks like the rigid bodies are moving now).

You can replace 
```go
m.Super().AddChild(instance)
fmt.Println(instance.GetPath())
m.boidsGD = append(m.boidsGD,instance.GetPath())
```
with
```go
m.boidsGD = append(m.boidsGD, m.Super().AsNode().AddChild(mob.AsNode()))
```

Big thanks for reporting your experience here, it really helps us to identify the pain points and areas to improve our Go integration with Godot!

**Reply by jugglingjsons** [reply]
Thanks, this approach indeed works.

However, I think it's suboptimal.
I would ideally aim for a way to manage and keep nodes in go memory (or expose an explicit possibility of doing that). 
This may come in handy when creating groupControllers or objectPooling.

I am not sure what the overhead of the `node := m.Super().GetNode(ctx, boidPath)` is, but I assume it would be faster if I could store and access the refs throughout the lifetime of the go process.

Context Scenario:
- simulations with thousands of actors

Would that be possible currenlty ? (I believe not)
Does it sound like a direction / approach the lib may take in the future?

Thanks for taking the time and very quick adjustment!

**Reply by Splizard** [reply]
I've thought about this, essentially it would be nice to have a safe set of Node methods, unfortunately node types are one of the least safe references to hold, as they aren't reference counted and may be freed unexpectedly when interacting with the scene tree.

If you are looking for better performance, you may have better luck working with the Server (ie `gd.RenderingServer`) singletons directly? Keep in mind, this project doesn't have any rendering/extensive benchmarks yet to measure overall performance compared to other languages working with Godot. Would love to see a contribution for this!

**Reply by jugglingjsons** [reply]
I've actually benchmarked it (the current implementation with casting and godot handling the drawing through nodes), half the CPU of the simulation was eaten by passing casting the nodes back and forth and passing the results of the calculation. I stand corrected :) 

Your suggestions is great, using servers directly is also what I've been hearing about when it came to spawning large numbers of entities with GDScript itself. Will give it a go and hopefully report back with a comparative example :)

**Reply by jugglingjsons** [reply]
Got it running relatively smooth.
I've played around with adjusting the pointer and casting in the lib itself to call draws in batches, but I don't know enough about it to be helpful.

If you are open to that I can clean up the code, and pr it as an example.

https://github.com/grow-graphics/gd/assets/20739064/79e29ae8-852a-4439-ae55-4d85c457759a

Depending on the radius & boid size the frames flactuate between 40-90 for 10k boids.

I wanted a proper example with multithreading. It's using a simple spacial partitioning as well.

**Reply by ShirenY** [reply]
> I've actually benchmarked it (the current implementation with casting and godot handling the drawing through nodes), half the CPU of the simulation was eaten by passing casting the nodes back and forth and passing the results of the calculation. I stand corrected :)
> 

This is really informative! Did you try comparing it to GDScript or C# implementations? As Go's FFI performance is a significant concern on the internet.

**Reply by jugglingjsons** [reply]
I am having way too much fun with this, but the original plan was to compare them.

I am using just the bare minimum of what I need to from Godot - so mostly the drawing engine. There's some overhead, but I am not sure if that's what slowing it down (I doubt it). From what I see half of the time is spent on the cgo stuff and half on loops and data copying.

That may not be true however, since I am using a lot of goroutines (both for drawing and for iteration)

That said, the biggest jumps I've seen were from:
- avoiding sqrt (so I guess math in SIMD would help more than optimising cgo)
- processing boids vs processing sequential grids
- data layout of structs

I am also using draw polygons  - instead of sprites drawing , which from what I know can be slower than sprites. Will report back once that changes.

One observation that is mostly true - drawing over 15k of anything slows things down significantly.
I can simulate up to 50k on 30 logic ticks with a differently sized map & grid. But I can't for the life of mine paint it and keep ok FPS.
I tried with Visibility toggle - will try again.

I may be using suboptimal strategy to synchronise data - my second largest overhead is channel send and channel receive.

My use case is very specific to swarms though, it may be different if you want to use Godot build ins a lot. I am building this to make my own local collision avoidance system though and this will be its exclusive use from my end.

Here's a more recent video:
There's: 
- shooting
- spatial partitioning 
- messed up fov
- dials for most of the stuff
- double buffers for most of stuff

https://github.com/grow-graphics/gd/assets/20739064/236947d5-a45a-4aba-8a3f-8aa45e221c74

I tried adjusting the drawing function a bit as well, like such (but I was going a bit blind, I don't understand everything about how it works):

```
func (self RenderingServer) CanvasItemSetTransforms2(items []gd.RID, transforms []gd.Transform2D) {
	var selfPtr = self.AsPointer()
	var ptr = mmm.API(selfPtr)

	var obj = self.AsObject()

	var frames []*callframe.Frame = []*callframe.Frame{}
	var argumentos []callframe.Args = []callframe.Args{}
	var uintptrs []uintptr = []uintptr{}
	for i, item := range items {
		var frame = callframe.New()
		callframe.Arg(frame, item)
		callframe.Arg(frame, transforms[i])
		frames = append(frames, frame)

		argumentos = append(argumentos, frame.Array(0))
		var r_ret callframe.Nil
		uintptrs = append(uintptrs, r_ret.Uintptr() )
		//ptr.Object.MethodBindPointerCall(ptr.Methods.RenderingServer.Bind_canvas_item_set_transform, obj, frame.Array(0), r_ret.Uintptr())
	}

	ptr.Object.MethodBindPointerCalls(ptr.Methods.RenderingServer.Bind_canvas_item_set_transform, obj, argumentos,uintptrs )

	for _, frame := range frames {
		frame.Free()
	}
}
```

I can upload this sample somewhere once I clean it up I guess.

As for cGo itself - read through this - https://shane.ai/posts/cgo-performance-in-go1.21/
I won't go in so much details myself.
If you'd like I can simplify my code to just random movement and see what it peaks at it terms of drawing :) 

EDIT: one more observation
- after adding: `"-ldflags=-s -w"` to the build command

 and 
```
	cmd.Env = append(os.Environ(), "CGO_CFLAGS=-O3 -march=native -mtune=native")
	cmd.Env = append(os.Environ(), "CGO_LDFLAGS=-O3 -flto")
```

to the env. I am seing a bump from 72-78 FPS to 80-85 FPS

**Reply by Splizard** [reply]
@jugglingjsons thanks for sharing, these are some fantastic performance insights! Are you using multi-mesh? It leverages batched rendering calls, so if you aren't already using it, it may increase rendering performance further (all sprites rendered with a single draw call). I guess it would depend on the overhead of interfacing with the underlying packed arrays used as the buffer.

**Reply by ShirenY** [reply]
Thanks, your swarms looks interesting!

As someone who has worked on fluid simulation with Godot, I suggest looking into compute shaders and manual instance drawing. However, I'm not sure how easy your algorithm can be parallelized.

**Reply by jugglingjsons** [reply]
Will try that as well, although I don't need more entities for what I am doing.
In the meantime I switched to textures - this was the biggest improvement since last update. Before that I have been drawing each boid & projectile as a regular polygon (which was way more taxing for the CPU).

I will look into multi mesh instance later on and perhaps for 3d, thanks for the pointers!
A huge thank you for your work on this project, I haven't had this much fun coding for ages :)
